controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -58,7 +60,7 @@ type KubeadmConfigGenerator interface {
 }
 
 type MachineGenerator interface {
-	GenerateMachine(ctx context.Context, c client.Client, namespace, namePrefix, clusterName, version string, infraRef, bootstrapRef *corev1.ObjectReference, labels map[string]string, owner *metav1.OwnerReference) error
+	GenerateMachine(ctx context.Context, input generator.MachineGeneratorInput) error
```

Comment:
I feel like this breaks the use of interfaces here by making the argument a type defined in a specific implementation. That said, I don't think it'd quite be right to define the input type here either...

I think the cleanest would be to make the input type an interface as well, but that seems like it's getting a bit meta...

@chuckha @ncdc what are your thoughts on where the most appropriate place to define the input type would be?

Reply:
I kinda want to get rid of the interfaces and find another approach to this 😄

Reply:
I'd agree on getting rid of interfaces if possible. Regarding the argument being a struct, it's a common pattern and it works fine in practice.

Reply:
I'm definitely open to ideas, as long as it doesn't involve standing up an apiserver and testing against it :joy: Having the ability to attach a debugger the tests and trace through the implementation has helped tremendously and I'd hate to lose that.

Reply:
I think I'd probably prefer a function pointer that is set by default when the reconciler is initialized (if it's nil), which allows us to inject test-specific behavior. I'll play around with it & see if I can come up w/anything I like.

Reply:
> Regarding the argument being a struct, it's a common pattern and it works fine in practice.

I'm not arguing that, more around where is the proper place to define the struct for the argument. Defining it with a concrete implementation doesn't seem right, but defining it with the interface doesn't necessarily feel right here either if we are considering the interface as something the consumer owns.

Reply:
I guess maybe we could define the interface in the source package, but IMO it's fine if the input is in the package you're using.

If we want to follow clusterctl, we could have `fake_<>.go` files for these structs